packages/microcosm/src/microcosm.js

```diff
@@ -423,21 +423,30 @@ class Microcosm extends Emitter {
     return snap.next !== next
   }
 
-  _updateSnapshotRange(source: Action, end: Action) {
-    let focus = source
+  _updateSnapshotRange(start: Action, stop: Action) {
+    let focus = start
     while (focus) {
       var changed = this._updateSnapshot(focus)
 
-      if (changed === false || focus === end) {
+      if (changed === false || focus === stop) {
         break
       }
 
       focus = focus.next
     }
   }
 
-  _removeSnapshot(action: Action) {
-    delete this.snapshots[action.id]
+  _removeSnapshotRange(start: Action, stop: ?Action) {
+    let focus = start
+    while (focus) {
+      delete this.snapshots[focus.id]
+
+      if (stop != null && focus !== stop) {
+        break
+      }
+
+      focus = focus.parent
+    }
```

Comment:
this is quite confusing. so it starts at an action, and moves up until it's passed the `stop`? unless there isn't a stop in which case it moves up until it hits the end of the chain?

Reply:
Yeah.

Reply:
so two things that feel weird here. 1 is that you're starting at a node and then moving up, where as `_updateSnapshotRange` (which seems like it should be very close in implementation) starts at a node and moves down. minor thing, but still a context switch to be made in reading both. 2. this `(stop != null && focus !== stop)`. negative switch statements are already tough, 2 of them compound.

Does History know what the range it wants to remove whenever it's emitting `remove`? If so, maybe we can change the arguments here to make things more clear here.

Reply:
something like this work (needs cleaning):
```js
_removeSnapshotRange(start: Action, stop: ?Action) {
  let focus = start
  
  if (stop == null) {
    delete this.snapshots[focus.id]
  } else {
    while (focus) {
      delete this.snapshots[focus.id]

      if (focus == stop) {
        break
      }
      
      focus = focus.next
    }
  }
}
```